markets/perps-market/contracts/storage/PerpsMarket.sol

```diff
@@ -96,77 +99,120 @@ library PerpsMarket {
+            latestLiquidationTimestamp != block.timestamp
         ) {
+            /**
+                if capacity is at 0, but the market is under configured liquidation p/d,
+                another block of liquidation becomes allowable.
+             */
             uint256 currentPd = MathUtil.abs(self.skew).divDecimal(marketConfig.skewScale);
             if (currentPd < maxLiquidationPd) {
-                liquidatableAmount = (
-                    requestedLiquidationAmount > maxAllowedLiquidationInWindow
-                        ? maxAllowedLiquidationInWindow
-                        : requestedLiquidationAmount
-                ).to128();
-                // track how much was utilized in this block to ensure any subsequent
-                // liquidations don't combine for larger than what's allotted in the window
-                self.lastUtilizedLiquidationCapacity = liquidatableAmount;
+                liquidatableAmount = MathUtil.min128(
+                    maxLiquidationInWindow.to128(),
+                    requestedLiquidationAmount
+                );
             }
         }
 
-        // only update timestamp if there is something being liquidated
         if (liquidatableAmount > 0) {
-            self.lastTimeLiquidationCapacityUpdated = block.timestamp.to128();
+            // if other liquidations happened in same block, add to amount otherwise set as liquidatable amt
+            self.liquidationAmounts[liquidationIndex] = self.liquidationTimestamps[
+                liquidationIndex
+            ] == block.timestamp
+                ? self.liquidationAmounts[liquidationIndex] + liquidatableAmount
+                : liquidatableAmount;
+            self.liquidationTimestamps[liquidationIndex] = block.timestamp.to128();
         }
     }
 
-    struct MarketUpdateData {
-        uint128 marketId;
-        int256 skew;
-        uint256 size;
-        int256 currentFundingRate;
-        int256 currentFundingVelocity;
+    /**
+     * @dev Returns the current liquidation capacity for the market
+     * @notice liquidationAmounts tracks every liquidation with the corresponding timestamp.
+     * The logic here goes through each item in the array and sums up the amounts that are within the
+     * configured liquidation window from the current block timestamp.  This value is used to determine
+     * the available capacity
+     */
+    function currentLiquidationCapacity(
+        Data storage self,
+        PerpsMarketConfiguration.Data storage marketConfig
+    )
+        internal
+        view
+        returns (uint capacity, uint256 maxLiquidationInWindow, uint128 latestLiquidationTimestamp)
+    {
+        uint accumulatedLiquidationAmounts;
+
+        uint maxConfiguredSeconds = marketConfig.maxSecondsInLiquidationWindow;
+        maxLiquidationInWindow = marketConfig.maxLiquidationAmountInWindow();
+
+        for (uint i = 0; i < maxConfiguredSeconds; i++) {
```

Comment:
Like i just want to avoid iterating over every second, like if the window is 30, it'll be 30 iterations... 
So what i suggest is setup an array, that appends whenever a liquidation happens or replaces (if many liquidations in the same block)
You can then iterate on it from end to beginign (i.e. latest liquidations to earliest) and exit when you're out of the window...
This avoid having to iterate over 30 seconds, if there are 4 records in the last 30 seconds, it'll just do 4 iterations

Reply:
we could create a new array that stores active indices and only access those values in the array.  seems like overkill tbh and plus storage costs of another array to track this.

Reply:
hmmm like you see, we might get into a situation where you have a guy with 5 markets getting liquidated simultaneously, and having to iterate 5 * 30 can become a bit cumbersome 🤔 
I was thinking of an array, of 1,000 placeholders, we keep adding to it, at a certain point, we exceed the 1,000, and the earliest get deleted.
In that array, we iterate from latest to earliest until we get out of the window... So it's similar to the design here, except you just iterate on per block cumulative liquidations